lib/mayaUsd/ufe/UsdAttributes.cpp

```diff
@@ -252,20 +188,61 @@ Ufe::Attribute::Ptr UsdAttributes::attribute(const std::string& name)
             ADD_UFE_USD_CTOR(Matrix3d),
             ADD_UFE_USD_CTOR(Matrix4d),
 #endif
+            { Ufe::Attribute::kString,
+              [](const UsdSceneItem::Ptr& si,
+                 UsdAttributeHolder*      attrHolder) -> Ufe::Attribute::Ptr {
+                  if (attrHolder->usdAttributeType() == PXR_NS::SdfValueTypeNames->String) {
+                      return UsdAttributeString::create(si, attrHolder);
+                  } else {
+                      return UsdAttributeToken::create(si, attrHolder);
+                  }
+              } },
+            { Ufe::Attribute::kEnumString,
+              [](const UsdSceneItem::Ptr& si,
+                 UsdAttributeHolder*      attrHolder) -> Ufe::Attribute::Ptr {
+                  if (attrHolder->usdAttributeType() == PXR_NS::SdfValueTypeNames->String) {
+                      return UsdAttributeEnumString::create(si, attrHolder);
+                  } else {
+                      return UsdAttributeEnumToken::create(si, attrHolder);
+                  }
+              } },
           };
-
 #undef ADD_UFE_USD_CTOR
-    auto                ctorIt = ctorMap.find(newAttrType);
+
     Ufe::Attribute::Ptr newAttr;
-    UFE_ASSERT_MSG(ctorIt != ctorMap.end(), kErrorMsgUnknown);
-#if defined(UFE_V4_FEATURES_AVAILABLE) && (UFE_PREVIEW_VERSION_NUM >= 4008)
-    if (ctorIt != ctorMap.end())
-        newAttr = ctorIt->second(fItem, fPrim, attributeDef, usdAttr);
-#else
-    if (ctorIt != ctorMap.end())
-        newAttr = ctorIt->second(fItem, usdAttr);
+
+#ifdef UFE_V4_FEATURES_AVAILABLE
+#if (UFE_PREVIEW_VERSION_NUM >= 4010)
+    // The shader definition always wins over a created attribute:
+    auto shaderPropAndType = _GetSdrPropertyAndType(fItem, name);
+    if (shaderPropAndType.first) {
+        auto ctorIt = ctorMap.find(usdTypeToUfe(shaderPropAndType.first));
+        UFE_ASSERT_MSG(ctorIt != ctorMap.end(), kErrorMsgUnknown);
+        if (ctorIt != ctorMap.end()) {
+            newAttr = ctorIt->second(
+                fItem,
+                new UsdShaderAttributeHolder(
+                    fPrim, shaderPropAndType.first, shaderPropAndType.second));
+        }
+    }
+#endif
 #endif
 
+    if (!newAttr) {
+        // No attribute for the input name was found -> create one.
+        PXR_NS::TfToken      tok(name);
+        PXR_NS::UsdAttribute usdAttr = _GetAttributeType(fPrim, name);
+        if (!usdAttr.IsValid()) {
+            return nullptr;
+        }
+        Ufe::Attribute::Type newAttrType = usdTypeToUfe(usdAttr);
+
+        auto ctorIt = ctorMap.find(newAttrType);
+        UFE_ASSERT_MSG(ctorIt != ctorMap.end(), kErrorMsgUnknown);
+        if (ctorIt != ctorMap.end())
+            newAttr = ctorIt->second(fItem, new UsdAttributeHolder(usdAttr));
```

Comment:
When dealing with only a USD attribute, only need the regular `UsdAttributeHolder`.

Reply:
As mentioned earlier, should be be using a smart pointer?